### PR TITLE
A-Level Compiler - WIP

### PR DESCRIPTION
**Pull request by brianpmaher**
# DO NOT MERGE

Push all A-Level work here.

### CODE REVIEW

File: parser.cs [hunk 1]
```diff
@@ -130,15 +131,17 @@ public class Parser {
...
                 break;
             default:
                 error(new List<TOKENS>{TOKENS.PROGRAM});
                 break;
         }
+        return progLabel;
     }
 
-    private void block() {
+    private void block(string startLabel, bool progBlock = false) {
```
Comment (JesseStephenBrown) [inline parser.cs]: This is unnecessary, we don't need to know that we are in progBlock. If you really want to know what to do in the semantic analyzer at genEnd, just look at the nesting level. Level 0 = main, so hlt. Level 1+ = not main, so ret.

File: SemAnalyzer.cs [hunk 2]
```diff
@@ -43,13 +42,17 @@ public class SemAnalyzer{
     }
 
     //Pops the table off the stack
-    public void genEnd() {
+    public void genEnd(bool endOfProg = false) {
```
Comment (JesseStephenBrown) [inline SemAnalyzer.cs]: This bool is unnecessary, check the nesting level.

File: parser.cs [hunk 3]
```diff
@@ -109,9 +109,9 @@ public class Parser {
         switch(__lookahead.Type) {
             case TOKENS.PROGRAM:
                 Console.Write(2 + " ");
-                programHeading();
+                string progLabel = programHeading();
                 match(TOKENS.SCOLON);
-                block();
+                block(progLabel, true);
```
Comment (JesseStephenBrown) [inline parser.cs]: block should only take in the label, not a boolean as well, see other comments

File: parser.cs (hunk 1 above)
Comment (brianpmaher) [inline parser.cs]: Good point.

File: SemAnalyzer.cs (hunk 2 above)
Comment (brianpmaher) [inline SemAnalyzer.cs]: ditto

File: parser.cs (hunk 3 above)
Comment (brianpmaher) [inline parser.cs]: ditto

File: SemAnalyzer.cs [hunk 4]
```diff
@@ -43,13 +42,17 @@ public class SemAnalyzer{
...
         SymbolTable top = SymbolTableStack.Peek();
         output(
             "SUB SP #" + top.Size + " SP",
-            "POP D" + top.NestingLevel,
-            "HLT"
+            "POP D" + top.NestingLevel
```
Comment (JesseStephenBrown) [inline SemAnalyzer.cs:49]: Also, did we need to pop in both cases?

Comment (brianpmaher) [inline SemAnalyzer.cs:49]: not sure, leaving until we know for sure.

File: testProgs/procedure.mp.asm [hunk 5]
```diff
@@ -0,0 +1,16 @@
+BR L0
+BR L1
```
Comment (JesseStephenBrown) [inline testProgs/procedure.mp.asm:2]: Is this generated? Because if so, it is wrong. We do not need to BR to the procedure. This line will never be hit.

**Pull request closed without merging**